### PR TITLE
Updating Stub Format for VMSMERouter Changes

### PR DESCRIPTION
**Pull request by mlarson02**
This PR updates the DTC emulation and Track Finding emulation Disk Stub data format by adding to the beginning of the stubword a bit determining whether the stub is in the negative z region of the detector, and adding an offset to the encoded r value for DiskPS stubs of 256 (7.5 cm) to allow fitting the negDisk bit in 36 bits.

This information is required in the AllStubs to be used for the new VMSMERouter dual FPGA module, and it was decided this was the easiest way to incorporate the information without making changes to the previous processing modules / causing the DTC/Input/AllStubs to be of different format. More information regarding this is available here.

PR also includes changes to written memory text files required for current master firmware-hls (writing offset r values for DiskPS stubs and negDisk bit for all disk stubs).

### CODE REVIEW

File: L1Trigger/TrackFindingTMTT/src/KFbase.cc [hunk 1]
```diff
@@ -551,7 +551,7 @@ namespace tmtt {
       matRinv = TMatrixD(TMatrixD::kInverted, matR);
     } else {
       // Protection against rare maths instability.
-      const TMatrixD unitMatrix(TMatrixD::kUnit, TMatrixD(nHelixPar_, nHelixPar_));
+      const TMatrixD unitMatrix(TMatrixD::kUnit, TMatrixD(2, 2));
```
Comment (aryd) [inline L1Trigger/TrackFindingTMTT/src/KFbase.cc:554]: Is this related to these changes?

File: L1Trigger/TrackFindingTracklet/README.md [hunk 2]
```diff
@@ -1,4 +1,4 @@
-To run the L1 tracking & create a TTree of tracking performance:
+To run the L1 tracking & create a TTree of tracking performance:
```
Comment (aryd) [inline L1Trigger/TrackFindingTracklet/README.md:1]: Remove extra space

File: L1Trigger/TrackFindingTracklet/interface/Settings.h [hunk 3]
```diff
@@ -344,7 +345,7 @@ namespace trklet {
 
     double kz() const { return 2.0 * zlength_ / (1 << nzbitsstub_[0]); }
     double kz(unsigned int layerdisk) const { return 2.0 * zlength_ / (1 << nzbitsstub_[layerdisk]); }
-    double kr() const { return rmaxdisk_ / (1 << nrbitsstub_[N_LAYER]); }
+    double kr() const { return rmaxdisk_ / (1 << (nrbitsstub_[N_LAYER] + 1)); }
```
Comment (aryd) [inline L1Trigger/TrackFindingTracklet/interface/Settings.h]: Can you add a comment to explain?

File: L1Trigger/TrackFindingTracklet/interface/Settings.h [hunk 4]
```diff
@@ -1051,18 +1053,23 @@ namespace trklet {
     double stripLength_PS_{0.1467};
     double stripLength_2S_{5.0250};
 
+    // The DR binning below disabled, as doesn't match latest FW.
+
     //Following values are used for duplicate removal
     //Rinv bins were optimised to ensure a similar number of tracks in each bin prior to DR
     //Rinv bin edges for 6 bins.
-    std::vector<double> rinvBins_{-rinvcut(), -0.004968, -0.003828, 0, 0.003828, 0.004968, rinvcut()};
+    //std::vector<double> rinvBins_{-rinvcut(), -0.004968, -0.003828, 0, 0.003828, 0.004968, rinvcut()};
```
Comment (aryd) [inline L1Trigger/TrackFindingTracklet/interface/Settings.h:1062]: Are these changes related to this updated?

File: L1Trigger/TrackFindingTracklet/interface/Tracklet.h [hunk 5]
```diff
@@ -221,6 +221,8 @@ namespace trklet {
 
     unsigned int PSseed() const { return ((layer() == 1) || (layer() == 2) || (disk() != 0)) ? 1 : 0; }
 
+    // Seed type:
```
Comment (aryd) [inline L1Trigger/TrackFindingTracklet/interface/Tracklet.h:224]: Why is this added?

File: L1Trigger/TrackFindingTracklet/src/Stub.cc [hunk 6]
```diff
@@ -160,7 +164,7 @@ double Stub::rapprox() const {
     else
       return settings_.rDSSouter(r_.value());
   }
-  return r_.value() * settings_.kr();
+  return (r_.value() + (1 << 8)) * settings_.kr();  // Incorporating r offset for disk PS stubs
```
Comment (aryd) [inline L1Trigger/TrackFindingTracklet/src/Stub.cc]: Can you do something like "int r_offset_bits=8" and use this variable?

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc [hunk 7]
```diff
@@ -403,9 +403,9 @@ void TrackletProcessor::execute(unsigned int iSector, double phimin, double phim
         if (negdisk) {
           indexz = ((1 << nbitszfinebintable_) - 1) - indexz;
         }
-        indexr = stub->r().value() >> (stub->r().nbits() - nbitsrfinebintable_);
+        indexr = stub->rvalue() >> (stub->r().nbits() + 1 - nbitsrfinebintable_);
```
Comment (aryd) [inline L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc]: Can you add comment about offset of +1

**Review by aryd — COMMENTED** [review]
I added a few questions and comments on the updates.

**Comment by mlarson02** [comment]
> I added a few questions and comments on the updates.

I've addressed the comments relating to my changes, the others are regarding code included in CMSSW 14 dev branch commits incorporated in my branch, but not yet merged into the ProjectionCalculator_without_IMATH_and_noncombined branch.
